### PR TITLE
Outlier gateway failure

### PR DESCRIPTION
**Pull request by JonathanO**
Signed-off-by: Jonathan Oddy <jonathan.oddy@transferwise.com>

*Description*:
Allows ejection of hosts for exceeding a number of consecutive 502,503,504 errors. This works similarly to the existing consecutive_5xx feature.

*API Changes*
https://github.com/envoyproxy/data-plane-api/pull/228

*Risk Level*: Medium

*Testing*:
Added unit test similar to the existing consecutive_5xx test.
Added unit test to check interaction between consecutive_gateway_error and consecutive_5xx ejection
Hasn't caught fire in a test environment yet.

### CODE REVIEW

File: test/common/upstream/outlier_detection_impl_test.cc [hunk 1]
```diff
@@ -133,15 +135,15 @@ TEST_F(OutlierDetectorImplTest, DestroyWithActive) {
       cluster_, empty_outlier_detection_, dispatcher_, runtime_, time_source_, event_logger_));
   detector->addChangedStateCb([&](HostSharedPtr host) -> void { checker_.check(host); });
 
-  loadRq(cluster_.hosts_[0], 4, 503);
+  loadRq(cluster_.hosts_[0], 4, 500);
```
Comment (htuch) [inline test/common/upstream/outlier_detection_impl_test.cc]: Don't we want to preserve the old behavior for 5xx (including 503) when `enforcingConsecutiveGatewayFailure()` is not enabled? Otherwise, we're making a breaking change here, which is subject to the deprecation policy https://github.com/envoyproxy/envoy/blob/master/DEPRECATED.md.

Comment (JonathanO) [reply test/common/upstream/outlier_detection_impl_test.cc]: Yes, and we do, however using a 503 in these tests will cause a logEject call (with enforced = false) for the ConsecutiveGatewayFailure as well. I felt that switching to 500 preserved what this test was trying to test, and left the test cleaner than if I'd added EXPECT_CALL for that additional call.

**Comment by JonathanO** [comment]
Now I think about it, there is a bit of this that leaves me slightly worried that I'm going to cause a visible change that's going to lead to some mild panic.
stats_.ejections_total_ is incremented even if the ejection isn't enforced.

Assuming the default config (consecutive_5xx = consecutive_gateway_failure = 5, and consecutive_gateway_failure not being enforced), then after 5 consecutive 503s there'll be an unenforced ConsecutiveGatewayFailure ejection and an enforced Consecutive5xx ejection. Both of these will bump the ejections_total_ counter! Anyone monitoring it will see double the value of ejections they saw before.

I'm not sure what to do about this.... it actually feels a bit wrong that the counter is incremented for an unenforced ejection, so perhaps it should only be incremented when we really do eject a host. If that's the case then what about the ejections_consecutive_5xx_ and ejections_consecutive_gateway_failure_ stats, should they also only be incremented if we really do eject a host?

**Comment by mattklein123** [comment]
@junr03 can you help @JonathanO with the questions/issues above?

**Comment by junr03** [comment]
The idea behind the individual counters is to gather information about the behavior of the outlier ejection subsystem, i.e the individual counters should increase even if an ejection was not enforced, because they inform if an ejection was _detected_. The ejection_active gauge is the only one that informs at any given time how many hosts are actually ejected. 

However, with your addition I agree that this becomes a little more confusing. I need to give a little more thought on what would be the most understandable way to approach this. I am thinking that your original desire is to have a more granular control of the codes that cause ejection, is that right? I wonder if there is desire to abstract this further so people can control subsets of error codes that cause ejection.

**Comment by JonathanO** [comment]
My aim was to be much more aggressive with ejecting for "network errors" (including those detected by other proxies) than for "application errors".

A similar problem to the one I've introduced arises already if, for example, I want to enforce/monitor success rate based detection, but not consecutive_5xx. Setting enforce to 0 for 5xx will still have it incrementing the ejections_total_ counter even though I have no interest in it. Of course, the individual detection method counters would be what you really want to track. Fortunately getting into this situation does require you to deviate from the defaults, so, while perhaps a surprise, it's probably OK.

I think my best option to preserve existing behaviour would be to add a config setting for each detection method to allow disabling it. A disabled detector would not increment any of the counters. The new detector would default to disabled.

I think it would also be useful to add counters for the enforced ejections, both total and broken out by detector. I know there's a gauge already for active ejections, but personal preference would be to have a total counter as well. I think the broken out counters would also be handy.

So, in short, I propose:
* Add config options to allow completely disabling detectors
* Add ejections_enforced_total counter
* Add ejections_enforced_consecutive_5xx (and gateway failure and success rate) counters.

I also wonder if the current naming is confusing: I certainly took the ejections counters to mean the number of times an ejection had actually happened! So I further propose:
* Deprecate ejections_consecutive_5xx and ejections_success_rate and replace with ejections_detected_consecutive_5xx and ejections_detected_success_rate
* Deprecate ejections_total and replace with ejections_detected_total

**Comment by htuch** [comment]
Would another option be to make `ejectHost()` only able to increment the counter once for a given host per request? I agree that having finer grained ejection counters would also be nice.

**Comment by JonathanO** [comment]
That would suppress the double counting in the situation where you have a completely default outlier detection config, however if you're running with e.g. consecutive_5xx=6 then after my change the counter would be incremented once at 5x503s (detected but not enforced) and again at 6 when the consecutive_5xx detector fires.

**Comment by junr03** [comment]
@JonathanO I like your proposal for stats. It makes the stats been ejected by this subsystem more clear, and observability more fine grained. 

I wonder, if after the stats breaking, there is still a need to provide the ability to completely turn off detectors.

**Comment by JonathanO** [comment]
I think I'd still need to avoid a change to the behaviour of the existing ejections_total_ counter even if I break the stats out. I could do that either with the disable/enable flag, or by not including the new detector in ejections_total_ (preserving current behaviour) which is probably OK since it'll be temporary if I'm going to deprecate it in order to rename it. The latter approach would work right now, but next time someone wants to add a new detector they'll run into this problem again.

**Comment by junr03** [comment]
Sorry, maybe I am being dense, but I thought your plan was to break out the stats between enforced and detected. If that is the case, and we deprecate the current total counter, I dont see why the disabled behavior is still necessary.

**Comment by JonathanO** [comment]
Sorry, I should have explained that a bit better.

Deprecating ejections_total in favour of ejections_detected_total solves my immediate problem. The new detector will only increment the new counter, preserving the behaviour of the old while we go through the deprecation cycle. However, next time someone adds a detector in a similar way they'll find that the new ejections_detected_total gets incremented. This will once again be a breaking change.

You're right that the disabled behaviour is unnecessary. The simplest solution by far is to not to have an ejections_detected_total counter at all, and leave summing the broken out by detector counters as an exercise for the user. I'll update the PR to do that and add the broken out counters.

**Comment by JonathanO** [comment]
In summary, the current state of the PR is that in addition to adding the consecutive_gateway_failure feature, which defaults to 0% enforcement, it adds the following counters:
* ejections_enforced_total: This counts the total number of enforced ejections
* ejections_detected_consecutive_5xx: Enforced + unenforced ejections for the consecutive_5xx detector
* ejections_enforced_consecutive_5xx: Enforced ejections for the consecutive_5xx detector
* ejections_detected_success_rate: Enforced + unenforced ejections for the success_rate detector
* ejections_enforced_success_rate: Enforced ejections for the success_rate detector
* ejections_detected_gateway_failure: Enforced + unenforced ejections for the gateway_failure detector
* ejections_enforced_gateway_failure: Enforced ejections for the gateway_failure detector

The _detected_ counters replace the existing ejections_consecutive_5xx and ejections_success_rate counters, which will be kept for a deprecation cycle.

The existing ejections_total counter will only count detection by the success_rate and consecutive_5xx detectors. This preserves current behaviour. I intend to mark this counter deprecated, and not replace it. If you need that number you will have to sum the individual _detected_ counters of interest. This avoids adding future detectors being a breaking change.

**Comment by mattklein123** [comment]
> The detected counters replace the existing ejections_consecutive_5xx and ejections_success_rate counters, which will be kept for a deprecation cycle. The existing ejections_total counter will only count detection by the success_rate and consecutive_5xx detectors. This preserves current behaviour. I intend to mark this counter deprecated, and not replace it. If you need that number you will have to sum the individual detected counters of interest. This avoids adding future detectors being a breaking change.

TBH I would skip deprecation and just make the changes. We can make dashboard changes at Lyft. It's not that big of a deal and I don't think we are at the point with counters where we have to be this strict. If anyone out there feels super strongly about this please speak up.

**Comment by junr03** [comment]
@JonathanO just to be clear adding a new detector would not be a breaking change for the ejections_detected_counter. The counter counts all ejections detected, so if you add a new detector that is detecting ejections those ejections _should_ be counted. But I think we are splitting hairs, and as you say the user can sum the detections they are interested in. 

About your other point in slack about enforcing ejections as a default. I think there can be a case made for all different types of outlier ejection to be opt in. But if you want to pursue that I would recommend opening another PR.

**Comment by JonathanO** [comment]
@mattklein123 While I'm happy to if you're sure there are no users of those counters outside of Lyft, I'm not sure that skipping the deprecation cycle buys us anything, as it doesn't simplify the change much. Since there is now a breaking change policy, it feels like we should stick to it unless there's good reason not to. If you'd still like me skip deprecation, I'll happily revise the PR tomorrow.

@junr03 Agreed, the change to default to 0% enforcement would be another PR.

**Comment by mattklein123** [comment]
Deprecation is fine if you don't mind. Can you merge master?

**Comment by JonathanO** [comment]
master is merged (6e5eb0b), but the Github hooks to reflect this in the PR don't seem to have fired. I'm going to do a one char change in the hope that it pleases Github.

File: source/common/upstream/outlier_detection_impl.cc [hunk 2]
```diff
@@ -186,15 +208,34 @@ bool DetectorImpl::enforceEjection(EjectionType type) {
...
+
 void DetectorImpl::ejectHost(HostSharedPtr host, EjectionType type) {
   uint64_t max_ejection_percent = std::min<uint64_t>(
       100, runtime_.snapshot().getInteger("outlier_detection.max_ejection_percent",
                                           config_.maxEjectionPercent()));
   double ejected_percent = 100.0 * stats_.ejections_active_.value() / host_monitors_.size();
   if (ejected_percent < max_ejection_percent) {
-    stats_.ejections_total_.inc();
+    if (type == EjectionType::Consecutive5xx || type == EjectionType::SuccessRate) {
+      // Deprecated counter, preserving old behaviour until it's removed.
```
Comment (htuch) [inline source/common/upstream/outlier_detection_impl.cc]: Can you add this to https://github.com/envoyproxy/envoy/blob/master/DEPRECATED.md?

File: source/common/upstream/outlier_detection_impl.h [hunk 3]
```diff
@@ -145,7 +147,14 @@ class DetectorHostMonitorImpl : public DetectorHostMonitor {
   GAUGE  (ejections_active)                                                                        \
   COUNTER(ejections_overflow)                                                                      \
   COUNTER(ejections_consecutive_5xx)                                                               \
-  COUNTER(ejections_success_rate)
+  COUNTER(ejections_success_rate)                                                                  \
+  COUNTER(ejections_enforced_total)                                                                \
```
Comment (htuch) [inline source/common/upstream/outlier_detection_impl.h]: Can you add these to https://github.com/envoyproxy/data-plane-api/blob/320b5afbd0f6ed88df8d8e361545decabfea7777/docs/root/configuration/cluster_manager/cluster_stats.rst (and also indicate there what is deprecated and not)?

File: source/common/upstream/outlier_detection_impl.cc [hunk 4]
```diff
@@ -229,33 +270,49 @@ void DetectorImpl::onConsecutive5xx(HostSharedPtr host) {
...
+    stats_.ejections_detected_consecutive_5xx_.inc();
+    ejectHost(host, EjectionType::Consecutive5xx);
+    host_monitors_[host]->resetConsecutive5xx();
+    break;
+  case EjectionType::ConsecutiveGatewayFailure:
+    stats_.ejections_detected_consecutive_gateway_failure_.inc();
+    ejectHost(host, EjectionType::ConsecutiveGatewayFailure);
+    host_monitors_[host]->resetConsecutiveGatewayFailure();
+    break;
+  default:
```
Comment (mattklein123) [inline source/common/upstream/outlier_detection_impl.cc]: nit: is this default case needed? Compiler should be OK with all cases implemented.

File: test/common/upstream/outlier_detection_impl_test.cc [hunk 5]
```diff
@@ -161,10 +163,11 @@ TEST_F(OutlierDetectorImplTest, DestroyHostInUse) {
 
   detector.reset();
 
-  loadRq(cluster_.hosts_[0], 5, 503);
+  loadRq(cluster_.hosts_[0], 5, 500);
 }
 
 TEST_F(OutlierDetectorImplTest, BasicFlow5xx) {
+
```
Comment (mattklein123) [inline test/common/upstream/outlier_detection_impl_test.cc]: new del newlines like these. Also, please add small comments on top of the new tests for what the test does (we are trying to do this for new tests).

File: source/common/upstream/outlier_detection_impl.h (hunk 3 above)
Comment (JonathanO) [reply source/common/upstream/outlier_detection_impl.h]: Done in https://github.com/envoyproxy/data-plane-api/pull/245/

File: source/common/upstream/outlier_detection_impl.cc (hunk 2 above)
Comment (JonathanO) [reply source/common/upstream/outlier_detection_impl.cc]: Done in f782640

File: test/common/upstream/outlier_detection_impl_test.cc [hunk 6]
```diff
@@ -176,17 +178,105 @@ TEST_F(OutlierDetectorImplTest, BasicFlow5xx) {
...
+  EXPECT_EQ(2UL, cluster_.info_->stats_store_.counter("outlier_detection.ejections_total").value());
+  EXPECT_EQ(
+      2UL,
+      cluster_.info_->stats_store_.counter("outlier_detection.ejections_consecutive_5xx").value());
+  EXPECT_EQ(0UL, cluster_.info_->stats_store_
+                     .counter("outlier_detection.ejections_consecutive_gateway_failure")
+                     .value());
+}
+
+/*
```
Comment (mattklein123) [inline test/common/upstream/outlier_detection_impl_test.cc]: nit: /** for beginning of block comments

File: source/common/upstream/outlier_detection_impl.cc [hunk 7]
```diff
@@ -229,33 +270,49 @@ void DetectorImpl::onConsecutive5xx(HostSharedPtr host) {
...
+    ejectHost(host, EjectionType::Consecutive5xx);
+    host_monitors_[host]->resetConsecutive5xx();
+    break;
+  case EjectionType::ConsecutiveGatewayFailure:
+    stats_.ejections_detected_consecutive_gateway_failure_.inc();
+    ejectHost(host, EjectionType::ConsecutiveGatewayFailure);
+    host_monitors_[host]->resetConsecutiveGatewayFailure();
+    break;
+  case EjectionType::SuccessRate:
+    break;
```
Comment (mattklein123) [inline source/common/upstream/outlier_detection_impl.cc]: Can you add a NOT_REACHED here (and replace break with it if the compiler accepts it)? This should never happen.
